src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -892,6 +892,114 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
     }
 }
 
+crate enum PreprocessingError<'a> {
+    Anchor(AnchorFailure),
+    Disambiguator(Range<usize>, String),
+    Resolution(ResolutionFailure<'a>, &'a str, Option<Disambiguator>),
+}
+
+impl From<AnchorFailure> for PreprocessingError<'_> {
+    fn from(err: AnchorFailure) -> Self {
+        Self::Anchor(err)
+    }
+}
+
+crate struct PreprocessingInfo<'a> {
+    crate path_str: &'a str,
+    disambiguator: Option<Disambiguator>,
+    extra_fragment: Option<&'a str>,
+    link_text: String,
+}
+
+/// Returns:
+/// - `None` if the link should be ignored.
+/// - `Some(Err)` if the link should emit an error
+/// - `Some(Ok)` if the link is valid
+///
+/// `link_buffer` is needed for lifetime reasons; it will always be overwritten and the contents ignored.
+crate fn preprocess_link<'a>(
+    ori_link: &'a MarkdownLink,
+    link_buffer_1: &'a mut String,
+    link_buffer_2: &'a mut String,
+) -> Option<Result<PreprocessingInfo<'a>, PreprocessingError<'a>>> {
+    // Bail early for real links.
+    if ori_link.link.contains('/') {
+        return None;
+    }
+
+    // [] is mostly likely not supposed to be a link
+    if ori_link.link.is_empty() {
+        return None;
+    }
+
+    *link_buffer_1 = ori_link.link.replace("`", "");
+    let mut parts = link_buffer_1.split('#');
+
+    let link = parts.next().unwrap();
+    if link.trim().is_empty() {
+        // This is an anchor to an element of the current page, nothing to do in here!
+        return None;
+    }
+    let extra_fragment = parts.next();
+    if parts.next().is_some() {
+        // A valid link can't have multiple #'s
+        return Some(Err(AnchorFailure::MultipleAnchors.into()));
+    }
+
+    // Parse and strip the disambiguator from the link, if present.
+    let no_backticks_range = range_between_backticks(&ori_link);
+    let (mut path_str, disambiguator) = match Disambiguator::from_str(&link) {
+        Ok(Some((d, path))) => (path.trim(), Some(d)),
+        Ok(None) => (link.trim(), None),
+        Err((err_msg, relative_range)) => {
+            if !should_ignore_link_with_disambiguators(link) {
+                // Only report error if we would not have ignored this link.
+                // See issue #83859.
+                let disambiguator_range = (no_backticks_range.start + relative_range.start)
+                    ..(no_backticks_range.start + relative_range.end);
+                return Some(Err(PreprocessingError::Disambiguator(disambiguator_range, err_msg)));
+            } else {
+                return None;
+            }
+        }
+    };
```

Comment:
`no_backticks_range` can be moved inside `if !should_ignore_link_with_disambiguators(link)`